In [ ]:

# ---
# Question: Predict if a person will have diabetes or not using the KNN algorithm.compare it to logistic regression
# ---

#### Trying to classify person as diabetes or not based on some attributes like Insulin, BloodPressure, Skin-Thickness, BMI, Age, Glucose, Pregnancies, DiabetesPedigreeFunction.

In [ ]:
import numpy as np
import pandas as pd
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score,roc_auc_score
from sklearn.preprocessing import MinMaxScaler
from sklearn.impute import SimpleImputer 
from sklearn.linear_model import LogisticRegression


In [ ]:

df = pd.read_csv('/content/drive/MyDrive/diabetes.csv')
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [ ]:
df.isnull().sum().sum()

0

In [ ]:
#Split, Fit and Transform the data

X= df[['Pregnancies', 'Glucose', 'BloodPressure','SkinThickness', 'Insulin', 'BMI',
       'DiabetesPedigreeFunction', 'Age']]
y=df['Outcome']

X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.2, random_state = 0)
#scaling

# X-train(fit,transform)
training_scaler = MinMaxScaler().fit(X_train)
training_set = training_scaler.transform(X_train)

#X_TEST(FIT,TRANSFORM)
testing_scaler = MinMaxScaler().fit(X_test)
testing_set = testing_scaler.transform(X_test)

In [ ]:
# Fit the Classifier
clf= KNeighborsClassifier(n_neighbors = 15)
clf.fit(training_set, y_train)
y_predict = clf.predict(testing_set)

In [ ]:
#Calculating accuracy of the model using classification_report and auc_Roc using KNN_classifier 
print("Accuracy:",metrics.accuracy_score(y_test, y_predict))
print(classification_report(y_test, y_predict))
print(roc_auc_score(y_test, y_predict))

Accuracy: 0.7467532467532467
              precision    recall  f1-score   support

           0       0.76      0.89      0.82        99
           1       0.71      0.49      0.58        55

    accuracy                           0.75       154
   macro avg       0.73      0.69      0.70       154
weighted avg       0.74      0.75      0.73       154

0.6898989898989898


### COMPARISONS WITH LOGISTIC REGRESSION

In [ ]:
#Calculating accuracy of the model using classification_report and auc_Roc using Logistic Regression
LogisticReg = LogisticRegression(max_iter = 300, random_state = 0)
LogisticReg.fit(X_train, y_train)
y_prediction = LogisticReg.predict(X_test)
print("Accuracy:",metrics.accuracy_score(y_test, y_prediction))
print(classification_report(y_test, y_prediction))
print(roc_auc_score(y_test, y_prediction))

Accuracy: 0.7467532467532467
              precision    recall  f1-score   support

           0       0.81      0.79      0.80        99
           1       0.64      0.67      0.65        55

    accuracy                           0.75       154
   macro avg       0.73      0.73      0.73       154
weighted avg       0.75      0.75      0.75       154

0.7303030303030302


**Observation**

Logistic regression gives a better prediction of those with diabetes as compared to KNN algorithm

Logistic Reg=73%

KNN=== 68%